src/blox/adam_utils.jl

```diff
@@ -0,0 +1,93 @@
+using ForwardDiff
+using LinearAlgebra
+
+function compute_jacs(sol, prob, tspan; dt=1)
+    jacs = Vector{Matrix}()
+    p = prob.p
+    
+    for _t in collect(tspan[1]:dt:tspan[2])
+        u0 = sol(_t)
+        function diff_function(u)
+            du = similar(u)
+            prob.f(du, u, p, _t)
+            du
+        end
+        push!(jacs, ForwardDiff.jacobian(diff_function, u0))
+    end
+
+    jacs
+end
+
+function lyap(jacs::Vector{Matrix}, dt=1)
+    T = size(jacs, 1)
+    n = size(jacs[1], 1)
+    q_store = I(n)
+    lexp = zeros(n)
+    lexp_counts = zeros(n)
+    for t ∈ 1:T
+        q, r = qr(jacs[t] * q_store)
+        d = sign.(diag(r))
+        d[d .== 0] .= 1
+        q_store = q*diagm(d)
```

Comment:
```suggestion
        q_store = q*Diagonal(d)
```
If you use `Diagonal` it's typically more efficient than `diagm` since it doesn't need to allocate the whole matrix